### ISSUE TITLE
tres compiler error in TDNSResolver::getResponse

### ISSUE THREAD

**Issue by mnordhoff** [ahupowerdns/hello-dns#38]
I'm on Ubuntu 16.04 with "gcc (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609".

Since 947908cb565cc4fb72517329a23b76ebd15eeeed `make tres` fails with this:

```
$ nice make tres
g++ -std=gnu++14 -Wall -O2 -MMD -MP -ggdb -Iext/simplesocket -Iext/simplesocket/ext/fmt-5.2.1/include -Iext/ -pthread    -c -o tres.o tres.cc
tres.cc: In member function ‘DNSMessageReader TDNSResolver::getResponse(const ComboAddress&, const DNSName&, const DNSType&, int)’:
tres.cc:129:32: error: converting to ‘const key_type {aka const std::tuple<ComboAddress, DNSName, DNSType>}’ from initializer list would use explicit constructor ‘constexpr std::tuple< <template-parameter-1-1> >::tuple(const _Elements& ...) [with _Elements = {ComboAddress, DNSName, DNSType}]’
   if(skips.count({server,dn,dt})) {
                                ^
tres.cc:185:36: error: converting to ‘std::set<std::tuple<ComboAddress, DNSName, DNSType> >::value_type {aka std::tuple<ComboAddress, DNSName, DNSType>}’ from initializer list would use explicit constructor ‘constexpr std::tuple< <template-parameter-1-1> >::tuple(const _Elements& ...) [with _Elements = {ComboAddress, DNSName, DNSType}]’
         skips.insert({server,dn,dt});
                                    ^
<builtin>: recipe for target 'tres.o' failed
make: *** [tres.o] Error 1
```

**Comment by ahupowerdns** [comment]
I will address this - some c++17-isms have crept in. For now, replace {server, dn, dt} by std::tie(server, dn, dt)

**Comment by ahupowerdns** [comment]
since travis can compile tdns again, this should be resolved. Can you check?

**Comment by mnordhoff** [comment]
Hmm. It compiles, but the output is mostly on one line. And it segfaults at the end.

```
$ ./tres mattnordhoff.net A
Retrieved . NSSET from hints, have 26 addresses
Result of query for mattnordhoff.net.|A (0 intermediate, 2 actual)
mattnordhoff.net. 300 A 104.18.56.175
mattnordhoff.net. 300 A 104.18.57.175
Used 6 queries
sh: 1: dot: not found
{"answer":[{"content":"104.18.56.175","name":"mattnordhoff.net.","ttl":300,"type":"A"},{"content":"104.18.57.175","name":"mattnordhoff.net.","ttl":300,"type":"A"}],"dot":"","intermediate":[],"msec":143,"name":"mattnordhoff.net.","numformerrs":0,"numqueries":6,"numtimeouts":0,"rcode":0,"trace":"mattnordhoff.net.|A Starting query at authority = ., have 26 addresses to try\nmattnordhoff.net.|A Sending to server k.root-servers.net. on 2001:7fd::1\nmattnordhoff.net.|A Received a 838 byte response with RCode Noerror, qname mattnordhoff.net., qtype A, aa: 0\nmattnordhoff.net.|A Authority net. IN NS 172800 c.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 m.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 d.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 f.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 k.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 e.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 a.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 i.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 h.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 g.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 l.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 b.gtld-servers.net.\nmattnordhoff.net.|A Authority net. IN NS 172800 j.gtld-servers.net.\nmattnordhoff.net.|A Additional a.gtld-servers.net. IN AAAA 172800 2001:503:a83e::2:30\nmattnordhoff.net.|A Additional b.gtld-servers.net. IN AAAA 172800 2001:503:231d::2:30\nmattnordhoff.net.|A Additional c.gtld-servers.net. IN AAAA 172800 2001:503:83eb::30\nmattnordhoff.net.|A Additional d.gtld-servers.net. IN AAAA 172800 2001:500:856e::30\nmattnordhoff.net.|A Additional e.gtld-servers.net. IN AAAA 172800 2001:502:1ca1::30\nmattnordhoff.net.|A Additional f.gtld-servers.net. IN AAAA 172800 2001:503:d414::30\nmattnordhoff.net.|A Additional g.gtld-servers.net. IN AAAA 172800 2001:503:eea3::30\nmattnordhoff.net.|A Additional h.gtld-servers.net. IN AAAA 172800 2001:502:8cc::30\nmattnordhoff.net.|A Additional i.gtld-servers.net. IN AAAA 172800 2001:503:39c1::30\nmattnordhoff.net.|A Additional j.gtld-servers.net. IN AAAA 172800 2001:502:7094::30\nmattnordhoff.net.|A Additional k.gtld-servers.net. IN AAAA 172800 2001:503:d2d::30\nmattnordhoff.net.|A Additional l.gtld-servers.net. IN AAAA 172800 2001:500:d937::30\nmattnordhoff.net.|A Additional m.gtld-servers.net. IN AAAA 172800 2001:501:b1f9::30\nmattnordhoff.net.|A Additional a.gtld-servers.net. IN A 172800 192.5.6.30\nmattnordhoff.net.|A Additional b.gtld-servers.net. IN A 172800 192.33.14.30\nmattnordhoff.net.|A Additional c.gtld-servers.net. IN A 172800 192.26.92.30\nmattnordhoff.net.|A Additional d.gtld-servers.net. IN A 172800 192.31.80.30\nmattnordhoff.net.|A Additional e.gtld-servers.net. IN A 172800 192.12.94.30\nmattnordhoff.net.|A Additional f.gtld-servers.net. IN A 172800 192.35.51.30\nmattnordhoff.net.|A Additional g.gtld-servers.net. IN A 172800 192.42.93.30\nmattnordhoff.net.|A Additional h.gtld-servers.net. IN A 172800 192.54.112.30\nmattnordhoff.net.|A Additional i.gtld-servers.net. IN A 172800 192.43.172.30\nmattnordhoff.net.|A Additional j.gtld-servers.net. IN A 172800 192.48.79.30\nmattnordhoff.net.|A Additional k.gtld-servers.net. IN A 172800 192.52.178.30\nmattnordhoff.net.|A Additional l.gtld-servers.net. IN A 172800 192.41.162.30\nmattnordhoff.net.|A Additional m.gtld-servers.net. IN A 172800 192.55.83.30\nmattnordhoff.net.|A Additional . IN OPT 0 \\# 0\nmattnordhoff.net.|A We got delegated to 13 net. nameserver names \nmattnordhoff.net.|A Have 26 IP addresses to iterate to: a.gtld-servers.net.=2001:503:a83e::2:30 a.gtld-servers.net.=192.5.6.30 b.gtld-servers.net.=2001:503:231d::2:30 b.gtld-servers.net.=192.33.14.30 c.gtld-servers.net.=2001:503:83eb::30 c.gtld-servers.net.=192.26.92.30 d.gtld-servers.net.=2001:500:856e::30 d.gtld-servers.net.=192.31.80.30 e.gtld-servers.net.=2001:502:1ca1::30 e.gtld-servers.net.=192.12.94.30 f.gtld-servers.net.=2001:503:d414::30 f.gtld-servers.net.=192.35.51.30 g.gtld-servers.net.=2001:503:eea3::30 g.gtld-servers.net.=192.42.93.30 h.gtld-servers.net.=2001:502:8cc::30 h.gtld-servers.net.=192.54.112.30 i.gtld-servers.net.=2001:503:39c1::30 i.gtld-servers.net.=192.43.172.30 j.gtld-servers.net.=2001:502:7094::30 j.gtld-servers.net.=192.48.79.30 k.gtld-servers.net.=2001:503:d2d::30 k.gtld-servers.net.=192.52.178.30 l.gtld-servers.net.=2001:500:d937::30 l.gtld-servers.net.=192.41.162.30 m.gtld-servers.net.=2001:501:b1f9::30 m.gtld-servers.net.=192.55.83.30 \n mattnordhoff.net.|A Starting query at authority = net., have 26 addresses to try\n mattnordhoff.net.|A Sending to server b.gtld-servers.net. on 192.33.14.30\n mattnordhoff.net.|A Received a 188 byte response with RCode Noerror, qname mattnordhoff.net., qtype A, aa: 0\n mattnordhoff.net.|A Authority mattnordhoff.net. IN NS 172800 cody.ns.cloudflare.com.\n mattnordhoff.net.|A Authority mattnordhoff.net. IN NS 172800 dana.ns.cloudflare.com.\n mattnordhoff.net.|A Additional cody.ns.cloudflare.com. IN A 172800 173.245.59.107\n mattnordhoff.net.|A Not accepting IP address of cody.ns.cloudflare.com.: out of authority of this server\n mattnordhoff.net.|A Additional cody.ns.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::adf5:3b6b\n mattnordhoff.net.|A Not accepting IP address of cody.ns.cloudflare.com.: out of authority of this server\n mattnordhoff.net.|A Additional dana.ns.cloudflare.com. IN A 172800 173.245.58.105\n mattnordhoff.net.|A Not accepting IP address of dana.ns.cloudflare.com.: out of authority of this server\n mattnordhoff.net.|A Additional dana.ns.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::adf5:3a69\n mattnordhoff.net.|A Not accepting IP address of dana.ns.cloudflare.com.: out of authority of this server\n mattnordhoff.net.|A Additional . IN OPT 0 \\# 0\n mattnordhoff.net.|A We got delegated to 2 mattnordhoff.net. nameserver names \n mattnordhoff.net.|A Don't have a resolved nameserver to ask anymore, trying to resolve 2 names\n mattnordhoff.net.|A Attempting to resolve NS dana.ns.cloudflare.com.|A\n  dana.ns.cloudflare.com.|A Starting query at authority = ., have 26 addresses to try\n  dana.ns.cloudflare.com.|A Sending to server f.root-servers.net. on 2001:500:2f::f\n  dana.ns.cloudflare.com.|A Received a 847 byte response with RCode Noerror, qname dana.ns.cloudflare.com., qtype A, aa: 0\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 l.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 b.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 c.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 d.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 e.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 f.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 g.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 a.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 h.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 i.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 j.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 k.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Authority com. IN NS 172800 m.gtld-servers.net.\n  dana.ns.cloudflare.com.|A Additional l.gtld-servers.net. IN A 172800 192.41.162.30\n  dana.ns.cloudflare.com.|A Additional l.gtld-servers.net. IN AAAA 172800 2001:500:d937::30\n  dana.ns.cloudflare.com.|A Additional b.gtld-servers.net. IN A 172800 192.33.14.30\n  dana.ns.cloudflare.com.|A Additional b.gtld-servers.net. IN AAAA 172800 2001:503:231d::2:30\n  dana.ns.cloudflare.com.|A Additional c.gtld-servers.net. IN A 172800 192.26.92.30\n  dana.ns.cloudflare.com.|A Additional c.gtld-servers.net. IN AAAA 172800 2001:503:83eb::30\n  dana.ns.cloudflare.com.|A Additional d.gtld-servers.net. IN A 172800 192.31.80.30\n  dana.ns.cloudflare.com.|A Additional d.gtld-servers.net. IN AAAA 172800 2001:500:856e::30\n  dana.ns.cloudflare.com.|A Additional e.gtld-servers.net. IN A 172800 192.12.94.30\n  dana.ns.cloudflare.com.|A Additional e.gtld-servers.net. IN AAAA 172800 2001:502:1ca1::30\n  dana.ns.cloudflare.com.|A Additional f.gtld-servers.net. IN A 172800 192.35.51.30\n  dana.ns.cloudflare.com.|A Additional f.gtld-servers.net. IN AAAA 172800 2001:503:d414::30\n  dana.ns.cloudflare.com.|A Additional g.gtld-servers.net. IN A 172800 192.42.93.30\n  dana.ns.cloudflare.com.|A Additional g.gtld-servers.net. IN AAAA 172800 2001:503:eea3::30\n  dana.ns.cloudflare.com.|A Additional a.gtld-servers.net. IN A 172800 192.5.6.30\n  dana.ns.cloudflare.com.|A Additional a.gtld-servers.net. IN AAAA 172800 2001:503:a83e::2:30\n  dana.ns.cloudflare.com.|A Additional h.gtld-servers.net. IN A 172800 192.54.112.30\n  dana.ns.cloudflare.com.|A Additional h.gtld-servers.net. IN AAAA 172800 2001:502:8cc::30\n  dana.ns.cloudflare.com.|A Additional i.gtld-servers.net. IN A 172800 192.43.172.30\n  dana.ns.cloudflare.com.|A Additional i.gtld-servers.net. IN AAAA 172800 2001:503:39c1::30\n  dana.ns.cloudflare.com.|A Additional j.gtld-servers.net. IN A 172800 192.48.79.30\n  dana.ns.cloudflare.com.|A Additional j.gtld-servers.net. IN AAAA 172800 2001:502:7094::30\n  dana.ns.cloudflare.com.|A Additional k.gtld-servers.net. IN A 172800 192.52.178.30\n  dana.ns.cloudflare.com.|A Additional k.gtld-servers.net. IN AAAA 172800 2001:503:d2d::30\n  dana.ns.cloudflare.com.|A Additional m.gtld-servers.net. IN A 172800 192.55.83.30\n  dana.ns.cloudflare.com.|A Additional m.gtld-servers.net. IN AAAA 172800 2001:501:b1f9::30\n  dana.ns.cloudflare.com.|A Additional . IN OPT 0 \\# 0\n  dana.ns.cloudflare.com.|A We got delegated to 13 com. nameserver names \n  dana.ns.cloudflare.com.|A Have 26 IP addresses to iterate to: a.gtld-servers.net.=192.5.6.30 a.gtld-servers.net.=2001:503:a83e::2:30 b.gtld-servers.net.=192.33.14.30 b.gtld-servers.net.=2001:503:231d::2:30 c.gtld-servers.net.=192.26.92.30 c.gtld-servers.net.=2001:503:83eb::30 d.gtld-servers.net.=192.31.80.30 d.gtld-servers.net.=2001:500:856e::30 e.gtld-servers.net.=192.12.94.30 e.gtld-servers.net.=2001:502:1ca1::30 f.gtld-servers.net.=192.35.51.30 f.gtld-servers.net.=2001:503:d414::30 g.gtld-servers.net.=192.42.93.30 g.gtld-servers.net.=2001:503:eea3::30 h.gtld-servers.net.=192.54.112.30 h.gtld-servers.net.=2001:502:8cc::30 i.gtld-servers.net.=192.43.172.30 i.gtld-servers.net.=2001:503:39c1::30 j.gtld-servers.net.=192.48.79.30 j.gtld-servers.net.=2001:502:7094::30 k.gtld-servers.net.=192.52.178.30 k.gtld-servers.net.=2001:503:d2d::30 l.gtld-servers.net.=192.41.162.30 l.gtld-servers.net.=2001:500:d937::30 m.gtld-servers.net.=192.55.83.30 m.gtld-servers.net.=2001:501:b1f9::30 \n   dana.ns.cloudflare.com.|A Starting query at authority = com., have 26 addresses to try\n   dana.ns.cloudflare.com.|A Sending to server b.gtld-servers.net. on 2001:503:231d::2:30\n   dana.ns.cloudflare.com.|A Received a 581 byte response with RCode Noerror, qname dana.ns.cloudflare.com., qtype A, aa: 0\n   dana.ns.cloudflare.com.|A Authority cloudflare.com. IN NS 172800 ns3.cloudflare.com.\n   dana.ns.cloudflare.com.|A Authority cloudflare.com. IN NS 172800 ns5.cloudflare.com.\n   dana.ns.cloudflare.com.|A Authority cloudflare.com. IN NS 172800 ns4.cloudflare.com.\n   dana.ns.cloudflare.com.|A Authority cloudflare.com. IN NS 172800 ns6.cloudflare.com.\n   dana.ns.cloudflare.com.|A Authority cloudflare.com. IN NS 172800 ns7.cloudflare.com.\n   dana.ns.cloudflare.com.|A Additional ns3.cloudflare.com. IN A 172800 162.159.0.33\n   dana.ns.cloudflare.com.|A Additional ns3.cloudflare.com. IN A 172800 162.159.7.226\n   dana.ns.cloudflare.com.|A Additional ns3.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:21\n   dana.ns.cloudflare.com.|A Additional ns3.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:7e2\n   dana.ns.cloudflare.com.|A Additional ns5.cloudflare.com. IN A 172800 162.159.2.9\n   dana.ns.cloudflare.com.|A Additional ns5.cloudflare.com. IN A 172800 162.159.9.55\n   dana.ns.cloudflare.com.|A Additional ns5.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:209\n   dana.ns.cloudflare.com.|A Additional ns5.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:937\n   dana.ns.cloudflare.com.|A Additional ns4.cloudflare.com. IN A 172800 162.159.1.33\n   dana.ns.cloudflare.com.|A Additional ns4.cloudflare.com. IN A 172800 162.159.8.55\n   dana.ns.cloudflare.com.|A Additional ns4.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:121\n   dana.ns.cloudflare.com.|A Additional ns4.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:837\n   dana.ns.cloudflare.com.|A Additional ns6.cloudflare.com. IN A 172800 162.159.3.11\n   dana.ns.cloudflare.com.|A Additional ns6.cloudflare.com. IN A 172800 162.159.5.6\n   dana.ns.cloudflare.com.|A Additional ns6.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:30b\n   dana.ns.cloudflare.com.|A Additional ns6.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:506\n   dana.ns.cloudflare.com.|A Additional ns7.cloudflare.com. IN A 172800 162.159.4.8\n   dana.ns.cloudflare.com.|A Additional ns7.cloudflare.com. IN A 172800 162.159.6.6\n   dana.ns.cloudflare.com.|A Additional ns7.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:408\n   dana.ns.cloudflare.com.|A Additional ns7.cloudflare.com. IN AAAA 172800 2400:cb00:2049:1::a29f:606\n   dana.ns.cloudflare.com.|A Additional . IN OPT 0 \\# 0\n   dana.ns.cloudflare.com.|A We got delegated to 5 cloudflare.com. nameserver names \n   dana.ns.cloudflare.com.|A Have 20 IP addresses to iterate to: ns3.cloudflare.com.=162.159.0.33 ns3.cloudflare.com.=162.159.7.226 ns3.cloudflare.com.=2400:cb00:2049:1::a29f:21 ns3.cloudflare.com.=2400:cb00:2049:1::a29f:7e2 ns4.cloudflare.com.=162.159.1.33 ns4.cloudflare.com.=162.159.8.55 ns4.cloudflare.com.=2400:cb00:2049:1::a29f:121 ns4.cloudflare.com.=2400:cb00:2049:1::a29f:837 ns5.cloudflare.com.=162.159.2.9 ns5.cloudflare.com.=162.159.9.55 ns5.cloudflare.com.=2400:cb00:2049:1::a29f:209 ns5.cloudflare.com.=2400:cb00:2049:1::a29f:937 ns6.cloudflare.com.=162.159.3.11 ns6.cloudflare.com.=162.159.5.6 ns6.cloudflare.com.=2400:cb00:2049:1::a29f:30b ns6.cloudflare.com.=2400:cb00:2049:1::a29f:506 ns7.cloudflare.com.=162.159.4.8 ns7.cloudflare.com.=162.159.6.6 ns7.cloudflare.com.=2400:cb00:2049:1::a29f:408 ns7.cloudflare.com.=2400:cb00:2049:1::a29f:606 \n    dana.ns.cloudflare.com.|A Starting query at authority = cloudflare.com., have 20 addresses to try\n    dana.ns.cloudflare.com.|A Sending to server ns6.cloudflare.com. on 2400:cb00:2049:1::a29f:506\n    dana.ns.cloudflare.com.|A Received a 67 byte response with RCode Noerror, qname dana.ns.cloudflare.com., qtype A, aa: 1\n    dana.ns.cloudflare.com.|A Answer says it is authoritative!\n    dana.ns.cloudflare.com.|A Answer dana.ns.cloudflare.com. IN A 86400 173.245.58.105\n    dana.ns.cloudflare.com.|A We got an answer to our question!\n    dana.ns.cloudflare.com.|A Additional . IN OPT 0 \\# 0\n    dana.ns.cloudflare.com.|A Done, returning 1 results, 0 intermediate\n mattnordhoff.net.|A Got 1 nameserver A addresses, adding to list\n mattnordhoff.net.|A We now have 1 resolved A addresses to try\n  mattnordhoff.net.|A Starting query at authority = mattnordhoff.net., have 1 addresses to try\n  mattnordhoff.net.|A Sending to server dana.ns.cloudflare.com. on 173.245.58.105\n  mattnordhoff.net.|A Received a 77 byte response with RCode Noerror, qname mattnordhoff.net., qtype A, aa: 1\n  mattnordhoff.net.|A Answer says it is authoritative!\n  mattnordhoff.net.|A Answer mattnordhoff.net. IN A 300 104.18.56.175\n  mattnordhoff.net.|A We got an answer to our question!\n  mattnordhoff.net.|A Answer mattnordhoff.net. IN A 300 104.18.57.175\n  mattnordhoff.net.|A We got an answer to our question!\n  mattnordhoff.net.|A Additional . IN OPT 0 \\# 0\n  mattnordhoff.net.|A Done, returning 2 results, 0 intermediate\n","type":"A"}
Segmentation fault
```

**Comment by ahupowerdns** [comment]
(this issue was fixed in e3891625e4c98c00cad97660a391104c73a767f8 )
